flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java

```diff
@@ -0,0 +1,264 @@
+    @GuardedBy("buffers")
+    private final Queue<MemorySegment> buffers = new ArrayDeque<>();
+
+    /** Whether this buffer pool has been destroyed or not. */
+    @GuardedBy("buffers")
+    private boolean destroyed;
+
+    /** Whether this buffer pool has been initialized or not. */
+    @GuardedBy("buffers")
+    private boolean initialized;
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize, long requestTimeout) {
+        checkArgument(totalBytes > 0, "Total memory size must be positive.");
+        checkArgument(bufferSize > 0, "Size of buffer must be positive.");
+        checkArgument(requestTimeout > 0, "Request timeout must be positive.");
+
+        this.totalBytes = totalBytes;
+        this.bufferSize = bufferSize;
+        this.requestTimeout = requestTimeout;
+
+        this.numTotalBuffers = (int) Math.min(totalBytes / bufferSize, Integer.MAX_VALUE);
+        this.numBuffersPerRequest =
+                Math.min(numTotalBuffers, Math.max(1, NUM_BYTES_PER_REQUEST / bufferSize));
+    }
+
+    @VisibleForTesting
+    long getTotalBytes() {
+        return totalBytes;
+    }
+
+    @VisibleForTesting
+    int getNumBuffersPerRequest() {
+        return numBuffersPerRequest;
+    }
+
+    @VisibleForTesting
+    int getNumTotalBuffers() {
+        return numTotalBuffers;
+    }
+
+    @VisibleForTesting
+    int getAvailableBuffers() {
+        synchronized (buffers) {
+            return buffers.size();
+        }
+    }
+
+    public int getMaxConcurrentRequests() {
+        return numBuffersPerRequest > 0 ? numTotalBuffers / numBuffersPerRequest : 0;
+    }
+
+    /** Initializes this buffer pool which allocates all the buffers. */
+    private void initialize() {
+        LOG.info(
+                "Initializing batch shuffle IO buffer pool: numBuffers={}, bufferSize={}.",
+                numTotalBuffers,
+                bufferSize);
+
+        checkState(
+                totalBytes >= bufferSize,
+                String.format(
+                        "Illegal configuration, config value for %s must be no smaller than %s,"
+                                + " please increase %s to at least %d bytes.",
+                        TaskManagerOptions.NETWORK_BATCH_SHUFFLE_READ_MEMORY.key(),
+                        TaskManagerOptions.MEMORY_SEGMENT_SIZE.key(),
+                        TaskManagerOptions.NETWORK_BATCH_SHUFFLE_READ_MEMORY.key(),
+                        bufferSize));
+
+        synchronized (buffers) {
+            checkState(!destroyed, "Buffer pool is already destroyed.");
+
+            if (initialized) {
+                return;
+            }
+            initialized = true;
+
+            try {
+                for (int i = 0; i < numTotalBuffers; ++i) {
+                    buffers.add(MemorySegmentFactory.allocateUnpooledOffHeapMemory(bufferSize));
+                }
+            } catch (OutOfMemoryError outOfMemoryError) {
```

Comment:
I think we might still have a call to `MemorySegment.free()` in case there might be future changes on how memorySegmentFactory allocates off-heap memory on OOM and destroy. We might have a method to cleanup the current `buffer` array to be called in the two places.